### PR TITLE
[Coupons] Restrictions: Allowed Emails Input and Data Handling

### LINKED ISSUE #5540: [Coupons] Restrictions: Add option to change the allowed emails for a coupon
I want to be able to change the allowed emails.
Screenshot 2021-12-27 at 12 23 46 PM

### PR DESCRIPTION
**Pull request by hafizrahman**
Closes: #5540

### Description

This PR adds and allows editing of allowed emails in the "Allowed Emails" screen. 

This PR also adds proper data passing and handling:
- Passting allowed emails data from Coupon Restriction screen to Allowed Emails screen,
- Passing merchant-edited emails input from Allowed Emails screen to Coupon Restriction screen.

### Testing instructions

1. In wp-admin **Marketing > Coupons**, create a Coupon with email restriction. To add the restriction, create a Coupon and go to **Usage Restrictions > Allowed emails**.
2. Go to the app, make sure Coupons beta toggle is enabled.
3. Open the details for the Coupon from step 1.
4. Click on the Menu, then click on **"Edit coupon"**, then click **"Usage Restrictions"**.
5. Tap **"Allowed emails"** button.
6. Make sure the list of allowed emails added on step 1 is shown correctly on the field.
7. Enter an invalid email then tap the X button. Make sure error message is shown.
8. Enter a wildcard email (e.g: *@woocommerce.com) then tap the X button. Make sure the screen returns properly to Coupon Restrictions screen.
9. Enter some valid values, go back to **Coupon Restrictions** screen, then go back to **Edit Coupon** screen and tap **Save** button. Make sure it is saved properly and returns to **Coupon Details** screen.
10. On the Coupon Details screen, in the **Coupon Summary** section, make sure the newly saved email restrictions are shown with the **"Restricted to customers with emails"** label.

Optionally go back and play around with adding multiple values, emptying all values, and ensure the data is consistent between the two screens.

### Video

https://user-images.githubusercontent.com/266376/172356395-ecf61e25-3d04-48eb-ad3d-5a9a43acbce5.mov

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionViewModel.kt [hunk 1]
```diff
@@ -20,26 +23,39 @@ class EmailRestrictionViewModel @Inject constructor(
...
-            ExitWithResult(viewState.allowedEmails)
+            val emails = viewState.allowedEmails.split(",").map { it.trim() }
+            val invalidEmails = emails.filter {
+                !StringUtils.isValidEmail(email = it, allowWildCardLocalPart = true)
+            }
+
+            if (invalidEmails.isEmpty()) {
+                ExitWithResult(viewState.allowedEmails)
+            } else {
+                ShowSnackbar(R.string.coupon_restrictions_allowed_emails_invalid)
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionViewModel.kt:50]: The error message in Core is smarter in that it tells merchants more specific issues (e.g: if the local part of email is wrong, or if it's the domain part, or if "@" is missing, and so on.

Here I just opt for a single error message for any invalidity issues. This is to match iOS behavior:

https://github.com/woocommerce/woocommerce-ios/blob/4b1db96e939af02bc33abcb3abe9956b05ba4b19/WooCommerce/Classes/ViewRelated/Coupons/Add%20and%20Edit%20Coupons/UsageDetails/CouponAllowedEmailsViewModel.swift#L24-L33

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionViewModel.kt [hunk 2]
```diff
@@ -20,26 +23,43 @@ class EmailRestrictionViewModel @Inject constructor(
...
+            // Return early as we allow emptying field
+            if (inputText.isEmpty()) {
+                ExitWithResult(inputText)
+            } else {
+                val emails = inputText.split(",").map { it.trim() }
+
+                if (emails.all { StringUtils.isValidEmail(email = it, allowWildCardLocalPart = true) }) {
+                    ExitWithResult(inputText)
+                } else {
+                    ShowSnackbar(R.string.coupon_restrictions_allowed_emails_invalid)
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionViewModel.kt:53]: The error message in Core is smarter in that it tells merchants more specific issues (e.g: whether the local part of email is wrong, or if it's the domain part, or if "@" is missing, and so on.)

Here I just opt for a single error message for any invalidity issues. This is to match iOS behavior:

https://github.com/woocommerce/woocommerce-ios/blob/4b1db96e939af02bc33abcb3abe9956b05ba4b19/WooCommerce/Classes/ViewRelated/Coupons/Add%20and%20Edit%20Coupons/UsageDetails/CouponAllowedEmailsViewModel.swift#L24-L33

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionViewModel.kt [hunk 3]
```diff
@@ -20,26 +23,46 @@ class EmailRestrictionViewModel @Inject constructor(
...
+                val emails = inputText.split(",").map { it.trim() }
+
+                // To match core, we want to allow:
+                // - regular emails, e.g: woo@woocommerce.com
+                // - emails with wildcard local part, e.g: *@woocommerce.com
+                if (emails.all { StringUtils.isValidEmail(email = it, allowWildCardLocalPart = true) }) {
+                    ExitWithResult(inputText)
+                } else {
+                    ShowSnackbar(R.string.coupon_restrictions_allowed_emails_invalid)
+                }
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionViewModel.kt:57]: The error message in Core is smarter in that it tells merchants more specific issues (e.g: whether the local part of email is wrong, or if it's the domain part, or if "@" is missing, and so on.)

Here I just opt for a single error message for any invalidity issues. This is to match iOS behavior:

https://github.com/woocommerce/woocommerce-ios/blob/4b1db96e939af02bc33abcb3abe9956b05ba4b19/WooCommerce/Classes/ViewRelated/Coupons/Add%20and%20Edit%20Coupons/UsageDetails/CouponAllowedEmailsViewModel.swift#L24-L33

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionScreen.kt [hunk 4]
```diff
@@ -1,9 +1,54 @@
...
+    ) {
+        WCOutlinedTextField(
+            value = viewState.allowedEmails,
+            label = stringResource(id = R.string.coupon_restrictions_allowed_emails),
+            onValueChange = onAllowedEmailsChanged,
+            helperText = stringResource(id = R.string.coupon_restrictions_allowed_emails_hint),
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(horizontal = dimensionResource(id = R.dimen.major_100))
+        )
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EmailRestrictionScreen.kt:52]: In Core, this is just a single line input field.

In the app, I opt to let the field to expand vertically if more emails are entered. I feel makes it easier to see and edit longer input (e.g: when merchants need to enter many different email addresses).

**Review by 0nko — APPROVED** [review]
The feature works great, nice job! 👍 

The only thing that seems a bit weird is the X button used for saving the input data. But this is a more general problem of the entire app since there are different approaches on different screens, so I don't want to block this PR.

**Pull request merged**